src/hotspot/share/runtime/stackOverflow.hpp

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_RUNTIME_STACKOVERFLOW_HPP
+#define SHARE_RUNTIME_STACKOVERFLOW_HPP
+
+#include "utilities/align.hpp"
+#include "utilities/debug.hpp"
+
+class JavaThread;
+
+// StackOverflow handling is encapsulated in this class.  This class contains state variables
+// for each JavaThread that implement stack overflow checking and guard page implementation.
+
+class StackOverflow {
+  friend class JVMCIVMStructs;
+  friend class JavaThread;
+ public:
+  // State of the stack guard pages for the containing thread.
+  enum StackGuardState {
+    stack_guard_unused,         // not needed
+    stack_guard_reserved_disabled,
+    stack_guard_yellow_reserved_disabled,// disabled (temporarily) after stack overflow
+    stack_guard_enabled         // enabled
+  };
+
+  StackOverflow() :
+    _stack_guard_state(stack_guard_unused),
+    _stack_overflow_limit(nullptr),
+    _reserved_stack_activation(nullptr),  // stack base not known yet
+    _stack_base(nullptr), _stack_end(nullptr) {}
+
+  // Initialization after thread is started.
+  void initialize(address base, address end) {
+     _stack_base = base;
+     _stack_end = end;
+    set_stack_overflow_limit();
+    set_reserved_stack_activation(base);
+  }
+ private:
+
+  StackGuardState  _stack_guard_state;
+
+  // Precompute the limit of the stack as used in stack overflow checks.
+  // We load it from here to simplify the stack overflow check in assembly.
+  address          _stack_overflow_limit;
+  address          _reserved_stack_activation;
+
+  // Support for stack overflow handling, copied down from thread.
+  address          _stack_base;
+  address          _stack_end;
+
+  address stack_end()  const           { return _stack_end; }
+  address stack_base() const           { assert(_stack_base != nullptr,"Sanity check"); return _stack_base; }
```

Comment:
We now keep stack  base and stack size/end both here and in Thread. Could we merge this and use this as a data holder for Thread?

Reply:
I thought about doing this but I'd rather carry the information than a pointer back to Thread.  It's only one extra word but doesn't invite abuse.

Reply:
Okay!